presto-hive-common/src/main/java/com/facebook/presto/hive/filesystem/ExtendedFileSystem.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.filesystem;
+
+import com.facebook.presto.hive.HiveFileContext;
+import com.facebook.presto.hive.HiveFileInfo;
+import com.facebook.presto.hive.NamenodeStats;
+import com.facebook.presto.hive.NestedDirectoryPolicy;
+import com.facebook.presto.hive.metastore.Table;
+import com.facebook.presto.hive.util.DirectoryLister;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.LocatedFileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.PathFilter;
+import org.apache.hadoop.fs.RemoteIterator;
+
+import java.io.IOException;
+import java.util.Iterator;
+
+public abstract class ExtendedFileSystem
+        extends FileSystem
+{
+    public abstract RemoteIterator<LocatedFileStatus> listDirectory(Path path)
+            throws IOException;
+
+    public abstract RemoteIterator<HiveFileInfo> listFiles(Path path)
+            throws IOException;
+
+    public abstract FSDataInputStream openFile(Path path, HiveFileContext hiveFileContext)
+            throws Exception;
+
+    public abstract Iterator<HiveFileInfo> list(
+            Table table,
+            Path path,
+            DirectoryLister directoryLister,
```

Comment:
It is little weird. 
I started this with binding DirectoryLister to concrete FS directly, but it was more messy.
The main reason was that we don't create the FileSystem through Injections, hence for the FS to use DirectoryLister, it would need to be passed to constructor directory.
This creates weird flows, especially for FileSystemWrapper.
Please note that in this case DIrectoryLister is just a utility providing specific functionality such as Timeout and caching.